custom_connectors/custom_auth/allocadia.rb

```diff
@@ -0,0 +1,1056 @@
+        }
+      end,
+
+      input_fields: lambda do |object_definitions|
+        object_definitions['budget'].only('budgetId').
+          concat([
+            {
+              name: 'columnNames',
+              type: :array,
+              of: :object,
+              properties: [{ name: 'columnName' }]
+            }
+          ]).concat(object_definitions['column'].only('location'))
+      end,
+
+      output_fields: lambda do |object_definitions|
+        [{
+           name: 'columns',
+           type: 'array',
+           of: 'object',
+           properties: object_definitions['column']
+        }]
+      end,
+
+    },
+  
+    add_choice: {
+      description: "Add <span class='provider'>column choice</span> " \
+        "in <span class='provider'>Allocadia</span>",
+
+      execute: lambda do |_connection, input|
+
+        param = '{ "label":"' + input['label'] + '", "externalAssociations": [{ "externalId": "' + input['externalId'] + '", "type": "CAMPAIGN" } ] }'
+
+        post("/v1/budgets/#{input['budgetId']}/columns/#{input['columnId']}/choices", parse_json(param))
+          .after_response do |code, body, headers|
+          # if /3\d{2} | 4\d{2} | 5\d{2}/.match?(code)
+          if code.to_s.match?(/[3-5]\d{2}/)
+            error("#{code}: #{body}")
+          else
+            { choiceId: headers['location'].split('/').last }
+          end
+        end
+
+      end,
+
+      input_fields: lambda do |object_definitions|
+        object_definitions['budget'].only('budgetId').concat(object_definitions['column'].only('columnId')).concat(object_definitions['choice'].ignored('choiceId'))
+      end,
+
+      output_fields: lambda do |object_definitions|
+        [{
+          name: 'choiceId',
+        }]
+      end,
+
+    },
+
+    get_item_by_id: {
+      description: "Get <span class='provider'>item by ID</span> " \
+        "in <span class='provider'>Allocadia</span>",
+
+     execute: lambda do |_connection, input|
+       
+        item = get("/v1/lineitems/#{input['itemId']}")
+        columns = call(:get_line_item_columns_raw, { budgetId: item['budgetId'] })
+
+        if input['primaryExternalColumnName']
+          primaryColId = columns.select {|col| col['name'] == input['primaryExternalColumnName'] }[0]['id']
+          choices = get("/v1/budgets/#{item['budgetId']}/columns/#{primaryColId}/choices")
+          if (item['cells'][primaryColId])
+            choiceId = item['cells'][primaryColId]['value']
+            choice = choices.select {|ch| ch['id'] == choiceId }[0]
+            externalId = choice['externalAssociations'].size > 0 ? choice['externalAssociations'][0]['externalId'] : nil
+            item['primaryExternalId'] = externalId
+          end
+        end
+        
+        cols = {}
+        columns.each do |col|
+          cols[col['id']] = col
+        end
+        item['cells'] = call(:format_item_cells, { columns: cols, cells: item['cells'] })
+       
+        item.except('_links')
```

Comment:
I have this here to avoid cluttering the job output with details that aren't useful. It can add a lot of extra noise, especially in the case for multiple line items.